File: test-orulo.js
// Script de teste para a integração com a API da Orulo
require('dotenv').config();
const axios = require('axios');

// Credenciais da Orulo
const CLIENT_ID = process.env.ORULO_CLIENT_ID || 'NVMeJ6c9nVZM3kMYhPBwU6CvL4CQbAgiZMLcuMgvfXA';
const CLIENT_SECRET = process.env.ORULO_CLIENT_SECRET || 'LFCtyXG3QP3V5o9pkXGjRJvKKf0gwozPiJ7TuDs8kmQ';

// URLs da API
const TOKEN_URL = 'https://www.orulo.com.br/oauth/token';
const BUILDINGS_URL = 'https://api.orulo.com.br/buildings';

// Chave da API (opcional, para tentar método alternativo)
const API_KEY = process.env.ORULO_API_KEY || '';

async function main() {
  try {
    // 1) Obter o token de autenticação
    console.log('Obtendo token OAuth...');
    
    const tokenParams = new URLSearchParams();
    tokenParams.append('grant_type', 'client_credentials');
    tokenParams.append('client_id', CLIENT_ID);
    tokenParams.append('client_secret', CLIENT_SECRET);
    
    const tokenResponse = await axios.post(TOKEN_URL, tokenParams, {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      }
    });
    
    const token = tokenResponse.data.access_token;
    console.log('Token obtido com sucesso!');
    console.log('- Token:', token.substring(0, 15) + '...');
    console.log('- Expira em:', tokenResponse.data.expires_in, 'segundos');
    
    // 2) Tentar diferentes métodos de autenticação
    console.log('\nTentando métodos alternativos de autenticação...');
    
    // Tentativa 1: Usando URL da documentação fornecida pelo usuário
    try {
      console.log('\nMétodo 1: Bearer token com URL da documentação');
      
      // Documentação da API da Orulo
      console.log('Informações para integração com a Orulo:');
      console.log('- Client ID:', CLIENT_ID);
      console.log('- Token obtido com sucesso:', token.substring(0, 15) + '...');
      
      // Tentando a URL exata fornecida pelo usuário
      const url1 = 'https://www.orulo.com.br/api/v2/buildings';
      console.log('URL:', url1);
      
      const authHeader = `Bearer ${token}`;
      console.log('Authorization:', authHeader);
      
      const response1 = await axios.get(url1, {
        headers: {
          'Authorization': authHeader,
          'Accept': 'application/json'
        }
      });
      
      console.log('Sucesso! Status:', response1.status);
      return response1.data;
    } catch (error1) {
      console.log('Falha no método 1:', error1.response?.status, error1.message);
      console.log('Detalhes:', error1.response?.data || {});
    }
    
    // Tentativa 2: API padrão com cliente_id e client_secret como parâmetros
    try {
      console.log('\nMétodo 2: client_id e client_secret como parâmetros');
      const url2 = 'https://api.orulo.com.br/buildings';
      console.log('URL:', url2);
      
      const response2 = await axios.get(url2, {
        params: { 
          page: 1, 
          per_page: 5,
          client_id: CLIENT_ID,
          client_secret: CLIENT_SECRET
        }
      });
      
      console.log('Sucesso! Status:', response2.status);
      return response2.data;
    } catch (error2) {
      console.log('Falha no método 2:', error2.response?.status, error2.message);
      console.log('Detalhes:', error2.response?.data || {});
    }
    
    // Tentativa 3: Usando X-Client-ID e X-Secret nos headers
    try {
      console.log('\nMétodo 3: X-Client-ID e X-Secret nos headers');
      const url3 = 'https://api.orulo.com.br/buildings';
      console.log('URL:', url3);
      
      const response3 = await axios.get(url3, {
        headers: {
          'X-Client-ID': CLIENT_ID,
          'X-Secret': CLIENT_SECRET,
          'Accept': 'application/json'
        },
        params: { page: 1, per_page: 5 }
      });
      
      console.log('Sucesso! Status:', response3.status);
      return response3.data;
    } catch (error3) {
      console.log('Falha no método 3:', error3.response?.status, error3.message);
      console.log('Detalhes:', error3.response?.data || {});
    }
    
    // Nenhum método funcionou
    throw new Error('Todos os métodos de autenticação falharam');
    
    // 3) Mostrar resultados
    const buildings = buildingsResponse.data;
    console.log('\n=== RESULTADO ===');
    console.log(`Total de imóveis encontrados: ${buildings.count || buildings.total || buildings.length || 0}`);
    
    if (buildings.results && buildings.results.length > 0) {
      console.log('\nPrimeiros imóveis:');
      buildings.results.slice(0, 3).forEach((building, index) => {
        console.log(`\n[${index + 1}] ${building.name || 'Sem nome'}`);
        console.log(`   - ID: ${building.id}`);
        console.log(`   - Endereço: ${building.address?.street || building.address || 'Não informado'}`);
        console.log(`   - Bairro: ${building.address?.neighborhood || building.neighborhood || 'Não informado'}`);
        console.log(`   - Preço: ${building.price_from || building.min_price || 'Não informado'}`);
      });
    }
    
    console.log('\nTeste concluído com sucesso!');
  } catch (error) {
    console.error('Erro:');
    
    if (error.response) {
      // A API retornou uma resposta com status de erro
      console.error(`Status: ${error.response.status}`);
      console.error('Dados:');
      console.error(JSON.stringify(error.response.data || error.response.statusText, null, 2));
    } else if (error.request) {
      // A requisição foi feita mas não houve resposta
      console.error('Sem resposta do servidor');
      console.error(error.request);
    } else {
      // Erro na configuração da requisição
      console.error('Erro na requisição:', error.message);
    }
  }
}

// Executar o teste
main();

File: test-orulo.py
import requests

# Credenciais da API Orulo
client_id = 'NVMeJ6c9nVZM3kMYhPBwU6CvL4CQbAgiZMLcuMgvfXA'
client_secret = 'LFCtyXG3QP3V5o9pkXGjRJvKKf0gwozPiJ7TuDs8kmQ'

# Método 1: Obter token pelo client credentials diretamente
token_url = 'https://www.orulo.com.br/oauth/token'

print('===== TESTE DA API ORULO =====')
print('Obtendo token OAuth via client credentials...')

# Solicitar token através do grant_type client_credentials
response = requests.post(token_url, data={
   'grant_type': 'client_credentials',
   'client_id': client_id,
   'client_secret': client_secret,
})

# Verifica se houve sucesso na request
if response.status_code != 200:
   print(f'Erro ao obter token: {response.status_code}')
   print(response.text)
   raise Exception('Falha ao gerar o token')

# Extrai o token de acesso do request
token_data = response.json()
access_token = token_data['access_token']
expires_in = token_data.get('expires_in', 'desconhecido')

print(f'Token obtido com sucesso!')
print(f'- Token: {access_token[:10]}...')
print(f'- Expira em: {expires_in} segundos')
print()

# Método 2: Usar o token que já temos (testando diferentes URLs)
# Token conhecido obtido anteriormente
known_token = 'ebLQ_HFn3RegIXU0D6Y7WfP1IHijjmw4heG8ZF7TuNg'

# Vamos testar diferentes endpoints
test_urls = [
    'https://www.orulo.com.br/api/v2/buildings',                # Todos os imóveis
    'https://www.orulo.com.br/api/v2/buildings?per_page=32',    # 32 imóveis por página
    'https://www.orulo.com.br/api/v2/buildings/ids/active',     # IDs ativos
]

print('===== TESTANDO DIFERENTES ENDPOINTS =====')
for url in test_urls:
    print(f'\nTentando acessar: {url}')
    
    # Teste com o token obtido agora
    response = requests.get(url, headers={
        'Authorization': f'Bearer {access_token}',
        'Accept': 'application/json',
    })
    
    if response.status_code == 200:
        data = response.json()
        
        # Verificar formato da resposta
        count = 0
        if isinstance(data, dict):
            if 'results' in data and isinstance(data['results'], list):
                count = len(data['results'])
                print(f'Sucesso! Formato padrão. Imóveis encontrados: {count}')
            elif 'buildings' in data and isinstance(data['buildings'], list):
                count = len(data['buildings'])
                print(f'Sucesso! Formato buildings. Imóveis encontrados: {count}')
            elif 'count' in data:
                count = data['count']
                print(f'Sucesso! Total no count: {count}')
            else:
                print(f'Sucesso! Formato desconhecido: {list(data.keys())}')
        elif isinstance(data, list):
            count = len(data)
            print(f'Sucesso! Lista direta. Itens encontrados: {count}')
        else:
            print(f'Sucesso! Formato desconhecido: {type(data)}')
            
        # Mostrar alguns detalhes se tiver imóveis
        if count > 0:
            if isinstance(data, dict):
                if 'results' in data and data['results']:
                    print(f"Primeiro imóvel: {data['results'][0].get('name', 'Sem nome')}")
                elif 'buildings' in data and data['buildings']:
                    print(f"Primeiro imóvel: {data['buildings'][0].get('name', 'Sem nome')}")
            elif isinstance(data, list) and data:
                print(f"Primeiro item: {data[0].get('name', 'Sem nome') if isinstance(data[0], dict) else data[0]}")
    else:
        print(f'Falha! Status: {response.status_code}')
        print(f'Resposta: {response.text[:200]}...' if len(response.text) > 200 else response.text)

print('\n===== TESTE DE BUSCA COM PARÂMETROS ESPECÍFICOS =====')
search_url = 'https://www.orulo.com.br/api/v2/buildings'
params = {
    'per_page': '32',
    'city': 'São Paulo',  # Testando filtro por cidade
}

print(f'Buscando imóveis com filtros: {params}')
response = requests.get(search_url, headers={
    'Authorization': f'Bearer {access_token}',
    'Accept': 'application/json',
}, params=params)

if response.status_code == 200:
    data = response.json()
    
    # Verificar formato da resposta
    count = 0
    if isinstance(data, dict):
        if 'results' in data and isinstance(data['results'], list):
            count = len(data['results'])
        elif 'buildings' in data and isinstance(data['buildings'], list):
            count = len(data['buildings'])
        elif 'count' in data:
            count = data['count']
    elif isinstance(data, list):
        count = len(data)
        
    print(f'Sucesso! Imóveis encontrados: {count}')
else:
    print(f'Falha! Status: {response.status_code}')
    print(f'Resposta: {response.text[:200]}...' if len(response.text) > 200 else response.text)

print('\n===== TESTE CONCLUÍDO =====')
